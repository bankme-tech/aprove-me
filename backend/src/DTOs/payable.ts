import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  IsDate,
  IsNotEmpty,
  IsNumber,
  IsString,
  IsUUID,
} from 'class-validator';

export class PayableDto {
  @ApiProperty({
    required: true,
    type: String,
    format: 'uuid',
    example: '2c91808f-e1d0-4b8b-b8c2-e8e7f7e8e8e8',
    description: 'ID do recebível',
  })
  @IsNotEmpty()
  @IsUUID()
  @IsString()
  id: string;

  @ApiProperty({
    required: true,
    type: Number,
    example: 1000,
    description: 'Valor do recebível',
  })
  @IsNotEmpty()
  @IsNumber()
  value: number;

  @ApiProperty({
    required: true,
    type: Date,
    example: '2021-06-01T00:00:00.000Z',
    description: 'Data de emissão do recebível',
  })
  @IsNotEmpty()
  @Type(() => Date)
  @IsDate()
  emissionDate: Date;

  @ApiProperty({
    required: true,
    type: String,
    format: 'uuid',
    example: '2c91808f-e1d0-4b8b-b8c2-e8e7f7e8e8e8',
    description: 'Referência do identificador do cedente',
  })
  @IsNotEmpty()
  @IsUUID()
  @IsString()
  assignor: string;
}
