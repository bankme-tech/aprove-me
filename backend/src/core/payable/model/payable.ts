import { Entity } from '@core/shared/model';
import { Notification } from '@core/shared/notification';
import { Id, PayableValue } from '@core/shared/value-objects';

export type PayableProps = {
  id?: string;
  value: number;
  emissionDate?: Date;
  assignor: string;
};

export class Payable extends Entity<PayableProps> {
  private _value: PayableValue;
  private _emissionDate?: Date;
  private _assignor: Id;

  constructor(props: PayableProps) {
    const notification = new Notification();
    super(props, notification);
    this._notification = notification;

    this._value = new PayableValue(props.value, this._notification);
    this._assignor = new Id(props.assignor, this._notification);
    this._emissionDate = props.emissionDate;
  }

  static create(props: PayableProps): Payable {
    return new Payable({
      id: props.id,
      value: props.value,
      assignor: props.assignor,
      emissionDate: props.emissionDate ?? new Date(),
    });
  }

  get id() {
    return this._id.getValue();
  }

  get value() {
    return this._value.getValue();
  }

  get assignor() {
    return this._assignor.getValue();
  }

  get emissionDate() {
    return this._emissionDate;
  }

  get notifications() {
    return this.getNotifications();
  }

  get containNotifications() {
    return this.hasNotifications();
  }
}
