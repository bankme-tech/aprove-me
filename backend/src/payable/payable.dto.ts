import {
  IsString,
  IsNotEmpty,
  Length,
  IsNumber,
  Min,
  ValidateNested,
  IsUUID,
} from 'class-validator';
import { Type } from 'class-transformer';

class CreateAssignorDto {
  @IsString()
  @IsNotEmpty()
  @Length(1, 140)
  name: string;

  @IsString()
  @IsNotEmpty()
  @Length(1, 30)
  document: string;

  @IsString()
  @IsNotEmpty()
  @Length(1, 140)
  email: string;

  @IsString()
  @IsNotEmpty()
  @Length(1, 20)
  phone: string;
}

export class CreatePayableDto {
  // @IsString()
  // @IsNotEmpty()
  // @Length(1, 100)
  // description: string;

  @IsNumber()
  @Min(0)
  @IsNotEmpty()
  amount: number;

  @IsString()
  @IsNotEmpty()
  emissionDate: string;
}

export class CreatePayableAssignorDto {
  @ValidateNested({ each: true })
  @Type(() => CreateAssignorDto)
  assignor: CreateAssignorDto | string;

  @ValidateNested({ each: true })
  @Type(() => CreatePayableDto)
  payables: CreatePayableDto[];
}
