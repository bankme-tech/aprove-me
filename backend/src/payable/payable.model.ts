export class Payable {
    constructor(
        public id: string,
        public value: number,
        public emissionDate: Date,
        public assignor: number
    ) { }
}