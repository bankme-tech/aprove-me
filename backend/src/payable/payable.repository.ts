import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { CreatePayableDto } from './dto/create-payable.dto';
import { UpdatePayableDto } from './dto/update-payable.dto';
import { Payable } from '@prisma/client';

@Injectable()
export class PayableRepository {
  constructor(private readonly prisma: PrismaService) {}

  async doesAssignorExist(assignorId: number): Promise<boolean> {
    const assignor = await this.prisma.assignor.findUnique({
      where: { id: assignorId },
    });

    return !!assignor;
  }

  async create(data: CreatePayableDto): Promise<Payable> {
    const { assignorId, ...payableData } = data;

    return this.prisma.payable.create({
      data: {
        ...payableData,
        assignor: { connect: { id: assignorId } },
      },
    });
  }

  async findAll(): Promise<Payable[]> {
    return this.prisma.payable.findMany();
  }

  async findOne(id: string): Promise<Payable | null> {
    return this.prisma.payable.findUnique({
      where: { id },
    });
  }

  async update(id: string, data: UpdatePayableDto): Promise<Payable> {
    const { assignorId, ...payableData } = data;

    return this.prisma.payable.update({
      where: { id },
      data: {
        ...payableData,
        assignor: assignorId ? { connect: { id: assignorId } } : undefined,
      },
    });
  }

  async remove(id: string): Promise<Payable> {
    return this.prisma.payable.delete({
      where: { id },
    });
  }
}
