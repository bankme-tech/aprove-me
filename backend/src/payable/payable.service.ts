import { Injectable } from '@nestjs/common';
import { CreatePayableDto } from './dto/create-payable.dto';
import { UpdatePayableDto } from './dto/update-payable.dto';
import PayableRepository from './repositories/payableRepository';

@Injectable()
export class PayableService {
  constructor(private readonly payableRepository: PayableRepository) {}

  async create(createPayableDto: CreatePayableDto) {
    return this.payableRepository.create(createPayableDto);
  }

  findAll() {
    return `This action returns all payable`;
  }

  findOne(id: number) {
    return `This action returns a #${id} payable`;
  }

  update(id: number, updatePayableDto: UpdatePayableDto) {
    return `This action updates a #${id} payable`;
  }

  remove(id: number) {
    return `This action removes a #${id} payable`;
  }
}
