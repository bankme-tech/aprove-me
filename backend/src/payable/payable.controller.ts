import { Controller, Get, Post, Body, Patch, Param, Delete, UseGuards } from '@nestjs/common';
import { PayableService } from './payable.service';
import { Prisma } from '@prisma/client';
import { CreatePayableAssignorDto } from './payable.dto';
import { AuthGuard } from 'src/auth/auth.guard';

@Controller('integrations/payable')
export class PayableController {
  constructor(private readonly payableService: PayableService) {}

  @Post()
  create(@Body() createPayableDto: CreatePayableAssignorDto) {
    return this.payableService.create(createPayableDto);
  }

  @UseGuards(AuthGuard)
  @Get()
  findAll() {
    return this.payableService.findAll();
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.payableService.findOne(id);
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() updatePayableDto: Prisma.PayableUpdateInput) {
    return this.payableService.update(id, updatePayableDto);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.payableService.remove(id);
  }
}
