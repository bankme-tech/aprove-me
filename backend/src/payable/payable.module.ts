import { Module } from '@nestjs/common';
import { PayableController } from './payable.controller';
import { CreatePayableUseCase } from './usecases/create-payable.usecase';
import { PrismaPayableRepository } from './repositories/prisma-payable.repository';
import { IPayableRepository } from './repositories/payable.repository.interface';
import { IPayableMapper } from './mappers/payable.mapper.interface';
import { PrismaPayableMapper } from './mappers/prisma-payable.mapper';
import { ICreatePayableUseCase } from './usecases/create-payable.usecase.interface';
import { AssignorModule } from 'src/assignor/assignor.module';
import { IFindAllPayablesUseCase } from './usecases/find-all-payables.usecase.interface';
import { FindAllPayablesUseCase } from './usecases/find-all-payables.usecase';
import { IFindPayableUseCase } from './usecases/find-payable.usecase.interface';
import { FindPayableUseCase } from './usecases/find-payable.usecase';
import { IUpdatePayableUseCase } from './usecases/update-payable.usecase.interface';
import { UpdatePayableUseCase } from './usecases/update-payable.usecase';
import { IRemovePayableUseCase } from './usecases/remove-payable.usecase.interface';
import { RemovePayableUseCase } from './usecases/remove-payable-usecase';
import { RabbitMQModule } from 'src/rabbitmq/rabbitmq.module';
import { RabbitMQProducer } from 'src/rabbitmq/rabbitmq.producer';
import { IProducer } from 'src/rabbitmq/interfaces/producer.interface';
import { EmailModule } from 'src/email/email.module';
import { PayableService } from './payable.service';

@Module({
  controllers: [PayableController],
  providers: [
    {
      provide: ICreatePayableUseCase,
      useClass: CreatePayableUseCase,
    },
    {
      provide: IFindAllPayablesUseCase,
      useClass: FindAllPayablesUseCase,
    },
    {
      provide: IPayableRepository,
      useClass: PrismaPayableRepository,
    },
    {
      provide: IFindPayableUseCase,
      useClass: FindPayableUseCase,
    },
    {
      provide: IUpdatePayableUseCase,
      useClass: UpdatePayableUseCase,
    },
    {
      provide: IRemovePayableUseCase,
      useClass: RemovePayableUseCase,
    },
    {
      provide: IPayableMapper,
      useClass: PrismaPayableMapper,
    },
    {
      provide: IProducer,
      useClass: RabbitMQProducer,
    },
    PayableService,
  ],
  imports: [AssignorModule, RabbitMQModule, EmailModule],
  exports: [PayableService],
})
export class PayableModule {}
