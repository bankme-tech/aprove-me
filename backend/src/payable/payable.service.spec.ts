import { Test, TestingModule } from '@nestjs/testing';
import { PayableService } from './payable.service';

describe('PayableService', () => {
  let service: PayableService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PayableService],
    }).compile();

    service = module.get<PayableService>(PayableService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
});
