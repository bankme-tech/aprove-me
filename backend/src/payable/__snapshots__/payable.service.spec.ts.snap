// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic CRUD operation (payable) should create two payables payable 1`] = `
{
  "amount": 1000,
  "description": "Please approve-me bank-me",
}
`;

exports[`basic CRUD operation (payable) should delete a payable 1`] = `
{
  "amount": 1000,
  "assignorId": "1",
  "description": "Test Payable",
  "id": "1",
}
`;

exports[`basic CRUD operation (payable) should return all playable 1`] = `
[
  {
    "amount": 10,
    "assignorId": "uuid3",
    "description": "example 1",
    "id": "uuid1",
  },
  {
    "amount": 150,
    "assignorId": "uuid3",
    "description": "example 2",
    "id": "uuid2",
  },
]
`;

exports[`basic CRUD operation (payable) should return an payable by id 1`] = `
{
  "amount": 1000,
  "assignorId": "1",
  "description": "Test Payable",
  "id": "1",
}
`;

exports[`basic CRUD operation (payable) should update an payable by id 1`] = `
{
  "amount": 1000,
  "assignId": "1",
  "description": "An updated description",
  "id": "1",
}
`;
